File: shell/gnome/src/prefs/accountDetail.js
import Adw from 'gi://Adw';
import Gtk from 'gi://Gtk';
import Pango from 'gi://Pango';
import { percentLeft } from '../format.js';
import { _ } from '../i18n.js';
import { accountName } from '../providers.js';
import {
    hiddenWindowsAfter,
    hiddenWindowsPatch,
    isStarred,
    isWindowHidden,
    starredAccountsPatch,
    starredAfter,
} from '../settings.js';
import { linksGroup, positionRow, removeRow, signInGroup, windowRow } from './accountDetailRows.js';
import { accountMark, canMove, detailSubtitle, noticeText } from './accountsModel.js';
import { providerImage } from './widgets.js';

const NOTICE_CLASSES = { signed_out: 'warning', no_subscription: 'warning', error: 'error' };

function label(cssClasses) {
    return new Gtk.Label({ xalign: 0, wrap: true, wrap_mode: Pango.WrapMode.WORD_CHAR, css_classes: cssClasses });
}

export class AccountDetail {
    constructor({ dir, client, account, actions }) {
        this._client = client;
        this._actions = actions;
        this._syncing = false;
        this._groups = [];
        this.id = account.id;
        this.widget = new Gtk.Box({ orientation: Gtk.Orientation.VERTICAL, spacing: 24 });
        this.widget.append(this._header(dir, account));
        this._identity = this._identityGroup();
        this.widget.append(this._identity);
    }

    update(account, context) {
        this._account = account;
        this._syncing = true;
        const shape = JSON.stringify(this._shapeOf(account, context));
        if (shape !== this._shape) this._rebuild(account, context, shape);
        this._syncHeader(account, context);
        this._syncIdentity(account, context);
        this._syncWindows(account, context.display);
        this._syncPosition(context);
        this._syncing = false;
    }

    _header(dir, account) {
        this._title = label(['title-3']);
        this._subtitle = label(['dim-label']);
        this._status = label([]);
        const texts = new Gtk.Box({ orientation: Gtk.Orientation.VERTICAL, spacing: 2, valign: Gtk.Align.CENTER });
        for (const child of [this._title, this._subtitle, this._status]) texts.append(child);
        const box = new Gtk.Box({ spacing: 12 });
        box.append(providerImage(dir, account.provider, 48));
        box.append(texts);
        return box;
    }

    _identityGroup() {
        this._visible = new Adw.SwitchRow({ title: _('Show in the panel and popup') });
        this._visible.connect('notify::active', () => this._guard(() => this._setHidden(!this._visible.active)));
        this._star = new Adw.SwitchRow({ title: _('Always show open') });
        this._star.connect('notify::active', () => this._guard(() => this._setStarred(this._star.active)));
        this._label = new Adw.EntryRow({ title: _('Label'), show_apply_button: true });
        this._label.connect('apply', () => this._client.setAccountLabel(this.id, this._label.text.trim()));
        const group = new Adw.PreferencesGroup();
        for (const row of [this._visible, this._star, this._label]) group.add(row);
        return group;
    }

    _shapeOf(account, context) {
        return [
            account.owner,
            account.windows.map(window => [window.id, window.label]),
            context.signIn !== null,
            context.links,
            context.order.length > 1,
        ];
    }

    _rebuild(account, context, shape) {
        for (const group of this._groups) this.widget.remove(group);
        this._windowRows = account.windows.map(window => windowRow(window, shown => this._setWindow(window.id, shown)));
        this._groups = [
            this._limitsGroup(),
            context.signIn ? signInGroup(account, () => this._actions.onSignIn(this._account)) : null,
            linksGroup(context.links, url => this._actions.onOpenLink(url)),
            this._manageGroup(account, context),
        ].filter(Boolean);
        for (const group of this._groups) this.widget.append(group);
        this._shape = shape;
    }

    _limitsGroup() {
        if (this._windowRows.length === 0) return null;
        const group = new Adw.PreferencesGroup({
            title: _('Limits'),
            description: _('Hidden limits leave the popup, the panel and notifications.'),
        });
        for (const row of this._windowRows) group.add(row);
        return group;
    }

    _manageGroup(account, context) {
        const group = new Adw.PreferencesGroup();
        this._position = context.order.length > 1 ? positionRow(delta => this._actions.onMove(this.id, delta)) : null;
        if (this._position) group.add(this._position.row);
        group.add(removeRow(account, () => this._actions.onRemove(this._account)));
        return group;
    }

    _syncHeader(account, context) {
        const mark = accountMark(account, context.display, context.offline);
        const status = mark.kind === 'notice' ? noticeText(account, mark.notice) : null;
        this._title.label = accountName(account);
        this._subtitle.label = detailSubtitle(account);
        this._status.label = status ?? '';
        this._status.visible = status !== null;
        this._status.css_classes = mark.kind === 'notice' ? [NOTICE_CLASSES[mark.notice]] : [];
    }

    _syncIdentity(account, context) {
        this._visible.active = !account.hidden;
        this._star.visible = context.canStar;
        this._star.active = isStarred(context.display, account.id);
        const editing = (this._label.get_state_flags() & Gtk.StateFlags.FOCUS_WITHIN) !== 0;
        if (!editing && this._label.text !== (account.label ?? '')) this._label.text = account.label ?? '';
    }

    _syncWindows(account, display) {
        account.windows.forEach((window, index) => {
            const row = this._windowRows[index];
            row.active = !isWindowHidden(display, account.id, window.id);
            row.subtitle = window.remainingPercent === null ? _('No data yet') : percentLeft(window.remainingPercent);
        });
    }

    _syncPosition(context) {
        if (!this._position) return;
        this._position.up.sensitive = canMove(context.order, this.id, -1);
        this._position.down.sensitive = canMove(context.order, this.id, 1);
    }

    _guard(run) {
        if (!this._syncing) run();
    }

    _setHidden(hidden) {
        this._client.setAccountHidden(this.id, hidden);
    }

    _setStarred(starred) {
        const display = this._client.settings?.display;
        if (display) this._client.updateSettings(starredAccountsPatch(starredAfter(display, this.id, starred)));
    }

    _setWindow(windowId, shown) {
        const display = this._client.settings?.display;
        if (this._syncing || !display) return;
        const windows = hiddenWindowsAfter(display, this.id, windowId, !shown);
        this._client.updateSettings(hiddenWindowsPatch(this.id, windows));
    }
}

File: crates/headroom-tray/src/ui/prefs/account_detail.rs
use std::cell::RefCell;
use std::rc::Rc;

use adw::prelude::*;

use super::account_actions::{Handlers, LinksGroup, ManageGroup, SignInGroup};
use super::account_list::provider_image;
use super::rows::SwitchRow;
use super::{Act, PrefsAction};
use crate::format::{percent_reading, window_label};
use crate::i18n::Lang;
use crate::payload::{Account, Display, ValueMode};
use crate::preferences::accounts::{LinkItem, Moves, offers_sign_in, sidebar_title};
use crate::preferences::change::Change;
use crate::preferences::choices::account_subtitle;

const HEADER_LOGO: i32 = 32;
const HEADER_SPACING: i32 = 12;

pub struct DetailInput<'a> {
    pub account: &'a Account,
    pub accounts: &'a [Account],
    pub display: &'a Display,
    pub moves: Moves,
    pub links: Vec<LinkItem>,
}

pub struct AccountDetail {
    pub page: adw::PreferencesPage,
    pub id: String,
    pub shape: Vec<(String, String)>,
    title: gtk::Label,
    subtitle: gtk::Label,
    visible: SwitchRow,
    star: SwitchRow,
    label: adw::EntryRow,
    windows: Vec<(String, SwitchRow)>,
    sign_in: SignInGroup,
    links: LinksGroup,
    manage: ManageGroup,
    display: Rc<RefCell<Display>>,
    account: Rc<RefCell<Account>>,
    lang: Lang,
}

#[must_use]
pub fn shape(account: &Account) -> Vec<(String, String)> {
    let head = (format!("{:?}", account.owner), account.provider.clone());
    std::iter::once(head)
        .chain(
            account
                .windows
                .iter()
                .map(|window| (window.id.clone(), window.label.clone())),
        )
        .collect()
}

fn header(account: &Account, logo_color: &str) -> (gtk::Box, gtk::Label, gtk::Label) {
    let row = gtk::Box::new(gtk::Orientation::Horizontal, HEADER_SPACING);
    row.append(&provider_image(&account.provider, logo_color, HEADER_LOGO));
    let texts = gtk::Box::new(gtk::Orientation::Vertical, 2);
    texts.set_valign(gtk::Align::Center);
    let title = gtk::Label::builder().xalign(0.0).wrap(true).build();
    title.add_css_class("title-3");
    let subtitle = gtk::Label::builder().xalign(0.0).wrap(true).build();
    subtitle.add_css_class("dim-label");
    subtitle.add_css_class("caption");
    texts.append(&title);
    texts.append(&subtitle);
    row.append(&texts);
    (row, title, subtitle)
}

fn visibility_row(lang: Lang, act: &Act, id: &str) -> SwitchRow {
    let (act, id) = (Rc::clone(act), id.to_owned());
    SwitchRow::new(lang.tr("Show in the tray and popup"), "", move |shown| {
        act(PrefsAction::SetHidden {
            account_id: id.clone(),
            hidden: !shown,
        });
    })
}

fn star_row(lang: Lang, act: &Act, id: &str, display: &Rc<RefCell<Display>>) -> SwitchRow {
    let (act, id, display) = (Rc::clone(act), id.to_owned(), Rc::clone(display));
    let row = SwitchRow::new(
        lang.tr("Always open in the popup"),
        lang.tr("Stays expanded when unstarred accounts collapse"),
        move |starred| {
            let ids = display.borrow().starred_after(&id, starred);
            act(PrefsAction::Change(Change::StarredAccounts(ids)));
        },
    );
    row.row
        .add_prefix(&gtk::Image::from_icon_name("starred-symbolic"));
    row
}

fn label_row(lang: Lang, act: &Act, id: &str) -> adw::EntryRow {
    let row = adw::EntryRow::builder()
        .title(lang.tr("Label"))
        .show_apply_button(true)
        .build();
    let (act, id) = (Rc::clone(act), id.to_owned());
    row.connect_apply(move |entry| {
        act(PrefsAction::SetLabel {
            account_id: id.clone(),
            label: entry.text().trim().to_owned(),
        });
    });
    row
}

fn window_row(act: &Act, id: &str, window: &str, display: &Rc<RefCell<Display>>) -> SwitchRow {
    let (act, id, window) = (Rc::clone(act), id.to_owned(), window.to_owned());
    let display = Rc::clone(display);
    SwitchRow::new("", "", move |shown| {
        let windows = display.borrow().hidden_windows_after(&id, &window, !shown);
        act(PrefsAction::Change(Change::HiddenWindows {
            account_id: id.clone(),
            windows,
        }));
    })
}

impl AccountDetail {
    pub fn new(lang: Lang, handlers: &Handlers, account: &Account, logo_color: &str) -> Self {
        let act = &handlers.act;
        let display: Rc<RefCell<Display>> = Rc::default();
        let shared = Rc::new(RefCell::new(account.clone()));
        let (head, title, subtitle) = header(account, logo_color);
        let detail = Self {
            page: adw::PreferencesPage::new(),
            id: account.id.clone(),
            shape: shape(account),
            title,
            subtitle,
            visible: visibility_row(lang, act, &account.id),
            star: star_row(lang, act, &account.id, &display),
            label: label_row(lang, act, &account.id),
            windows: account
                .windows
                .iter()
                .map(|window| {
                    let row = window_row(act, &account.id, &window.id, &display);
                    (window.id.clone(), row)
                })
                .collect(),
            sign_in: SignInGroup::new(lang, handlers, &shared),
            links: LinksGroup::new(lang, act),
            manage: ManageGroup::new(lang, handlers, &shared),
            display,
            account: shared,
            lang,
        };
        detail.assemble(&head);
        detail
    }

    fn assemble(&self, head: &gtk::Box) {
        let general = adw::PreferencesGroup::new();
        general.add(head);
        let rows = adw::PreferencesGroup::new();
        rows.add(&self.visible.row);
        rows.add(&self.star.row);
        rows.add(&self.label);
        self.page.add(&general);
        self.page.add(&rows);
        if !self.windows.is_empty() {
            let limits = adw::PreferencesGroup::builder()
                .title(self.lang.tr("Limits"))
                .description(
                    self.lang
                        .tr("Hidden limits leave the popup, the tray and notifications."),
                )
                .build();
            for (_, row) in &self.windows {
                limits.add(&row.row);
            }
            self.page.add(&limits);
        }
        self.page.add(&self.sign_in.group);
        self.page.add(&self.links.group);
        self.page.add(&self.manage.group);
    }

    pub fn update(&self, input: DetailInput) {
        let (account, display) = (input.account, input.display);
        let title = sidebar_title(account, input.accounts);
        self.title.set_text(&title);
        self.subtitle
            .set_text(&account_subtitle(self.lang, account));
        self.visible.set(!account.hidden);
        self.star.set(display.is_starred(&account.id));
        self.update_label(account);
        for (window_id, row) in &self.windows {
            if let Some(window) = account
                .windows
                .iter()
                .find(|window| window.id == *window_id)
            {
                row.row
                    .set_title(&window_label(self.lang, &window.id, &window.label));
                row.row.set_subtitle(&percent_reading(
                    self.lang,
                    window.remaining_percent,
                    ValueMode::Left,
                ));
            }
            row.set(!display.is_window_hidden(&account.id, window_id));
        }
        self.sign_in.update(&title, offers_sign_in(account));
        self.links.update(input.links);
        self.manage.update(input.moves);
        *self.display.borrow_mut() = display.clone();
        *self.account.borrow_mut() = account.clone();
    }

    fn update_label(&self, account: &Account) {
        let editing = self
            .label
            .state_flags()
            .contains(gtk::StateFlags::FOCUS_WITHIN);
        let label = account.label.as_deref().unwrap_or_default();
        if !editing && self.label.text() != label {
            self.label.set_text(label);
        }
    }
}
